File: core/src/main/java/com/ataulm/chunks/Items.java
package com.ataulm.chunks;

import com.google.auto.value.AutoValue;

import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import javax.annotation.Nullable;

@AutoValue
public abstract class Items implements Iterable<Item> {

    public static Items create(List<Item> entries) {
        return new AutoValue_Items(Collections.unmodifiableList(entries));
    }

    public static Items empty() {
        return create(Collections.<Item>emptyList());
    }

    public abstract List<Item> entries();

    protected Items() {
        // use static factory
    }

    public boolean isEmpty() {
        return entries().isEmpty();
    }

    public Item get(int position) {
        return entries().get(position);
    }

    public boolean containsEntryWith(Id id) {
        return findEntryWith(id) != null;
    }

    @Nullable
    public Item findEntryWith(Id id) {
        for (Item item : entries()) {
            if (item.id().equals(id)) {
                return item;
            }
        }
        return null;
    }

    public int size() {
        return entries().size();
    }

    @Override
    public Iterator<Item> iterator() {
        return entries().iterator();
    }

}

File: core/src/main/java/com/ataulm/chunks/Items.kt
package com.ataulm.chunks

import java.util.*

data class Items(val entries: List<Item>)
    : Iterable<Item> {

    val isEmpty: Boolean
        get() = entries.isEmpty()

    operator fun get(position: Int): Item {
        return entries[position]
    }

    fun containsEntryWith(id: Id): Boolean {
        return findEntryWith(id) != null
    }

    fun findEntryWith(id: Id): Item? {
        for (item in entries) {
            if (item.id == id) {
                return item
            }
        }
        return null
    }

    fun size(): Int {
        return entries.size
    }

    override fun iterator(): Iterator<Item> {
        return entries.iterator()
    }

    companion object {

        fun create(entries: List<Item>): Items {
            return Items(Collections.unmodifiableList(entries))
        }

        fun empty(): Items {
            return create(emptyList())
        }
    }

}

